### PR TITLE
Fix pfor defunc issue

### PR DESCRIPTION
**Pull request by bhack**
This is trying to fix the indirect defunc case with a vectorized_map

### CODE REVIEW

**Comment by bhack** [comment]
@mdanatg  I've tried to add a dummy failing test (we could improve it as you like).

The goal here is to fix https://github.com/tensorflow/tensorflow/issues/55340

I see that we are checking if we are in eager mode and we wrap the function in that case.

https://github.com/tensorflow/tensorflow/blob/f62c8863b3e19b81d2484a95e96ddb854ea66b2f/tensorflow/python/ops/parallel_for/control_flow_ops.py#L183-L205

/cc @wangpengmit

**Comment by bhack** [comment]
@mdanatg I've tried to insert  `f = autograph.tf_convert(f, autograph_ctx.control_status_ctx())` there but it is failing with the same error

**Comment by bhack** [comment]
I've moved `autograph.tf_convert` internally. I don't know if the coverage is complete but locally the test is passing.
Let me know if and how we want to improve this.

**Comment by bhack** [comment]
@gbaned Can you connect this PR to https://github.com/tensorflow/tensorflow/issues/55340. So when it is merged it will close the ticket.

**Review by mdanatg — APPROVED** [review]
Looks good overall.

File: tensorflow/python/ops/parallel_for/control_flow_ops_test.py [hunk 1]
```diff
@@ -97,6 +97,16 @@ def loop_fn(i):
       self._test_loop_fn(loop_fn, 3, fallback_to_while_loop=False)
     self._test_loop_fn(loop_fn, 3, fallback_to_while_loop=True)
 
+  def test_defun(self):        
+    def_function.function
```
Comment (mdanatg) [inline tensorflow/python/ops/parallel_for/control_flow_ops_test.py:101]: Forgot the `@` (for the decorator)

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/parallel_for/control_flow_ops_test.py (hunk 1 above)
Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py]: Yes it was a typo but now I've removed the function here and keep it only in the parent function as  I think probably it is a better test.

**Review by mdanatg — APPROVED** [review]

File: tensorflow/python/ops/parallel_for/control_flow_ops_test.py (hunk 1 above)
Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py]: Now it is totally failing (no session etc..). Do we need to limit this test to graph mode only?

**Comment by bhack** [comment]
@mdanatg Let me know if you like this test or you want to test something else.

File: tensorflow/python/ops/parallel_for/control_flow_ops_test.py [hunk 2]
```diff
@@ -2654,6 +2654,29 @@ def loop_fn(i):
...
+      return math_ops.reduce_sum(inner(y)) + 2
+
+    z = random_ops.random_uniform([4, 2])
+    
+    def loop_fn(i):
+      return outer(array_ops.gather(z, i))
+
+    @def_function.function
+    def loop_fn_caller():
+      self._test_loop_fn(loop_fn, 4)
```
Comment (mdanatg) [inline tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2675]: I don't think _test_loop_fn works inside tf.function.

Comment (mdanatg) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2675]: You probably just need to remove the decorator, and simplify loop_fn to just contain an autograph loop, no other inner/outer functions.

Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: I removed inner and outer that was just a copy of the test we had below.

> I don't think _test_loop_fn works inside tf.function.

Isn't exactly this what we want to fix and what was failing in the example we collected in the ticket https://github.com/tensorflow/tensorflow/issues/55340#issue-1177342161 ?

Comment (mdanatg) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: I think it is, but `_test_loop_fn` itself contains test-only code like `self.evaluate`, that won't work the way you expect. So you might need to write a test in a way which doesn't use it, by manually calling whatever `_test_loop_fn` called.

Comment (mdanatg) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: It might work though (accidentally). Let's see what the tests say.

**Review by mdanatg — APPROVED** [review]

File: tensorflow/python/ops/parallel_for/control_flow_ops_test.py (hunk 2 above)
Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: Ok tests are done. What do u think?

Comment (mdanatg) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: Looking good, but I would add a TODO / caution in that test - I'm not sure if it works. If you still have the local build, try seeing if the test fails without the fix.

Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: > If you still have the local build,

I had tested it but now the cache is invalid.
I made an extra commit here to bring it back to its original behavior so you can see yourself abusing the CI a bit.

I have only a doubt about the codepath coverage as we still have two others conditional `loop_fn(` calls in the same file.

Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: Please check the test log, so I will revert the commit when you have done.

Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2667]: You can check the log.
It Is failing as in ticket: `TypeError: 'Tensor' object cannot be interpreted as an integer`.
Let me know I will revert the commit

Comment (bhack) [reply tensorflow/python/ops/parallel_for/control_flow_ops_test.py:2675]: Reverted. You can check the failing log at https://source.cloud.google.com/results/invocations/9e2f7686-e850-43e2-92eb-a3648f822ab4/targets/%2F%2Ftensorflow%2Fpython%2Fops%2Fparallel_for:control_flow_ops_test_cpu/tests

**Review by mdanatg — APPROVED** [review]

**Pull request merged**